lib/cluster/election.c

```diff
@@ -496,59 +573,30 @@ election_count_vote(election_t *e, xmlNode *vote, bool can_win)
         log_level = LOG_WARNING;
         done = TRUE;
 
-    } else if (crm_str_eq(op, CRM_OP_NOVOTE, TRUE)) {
-        char *op_copy = NULL;
-        char *uname_copy = NULL;
-
+    } else if (crm_str_eq(vote.op, CRM_OP_NOVOTE, TRUE)
+               || crm_str_eq(vote.from, e->uname, TRUE)) {
+        /* Receiving our own broadcast vote, or a no-vote from peer, is a vote
+         * for us to win
+         */
         if (!we_are_owner) {
             crm_warn("Cannot count %s %s from %s because we are not election owner (%s)",
-                     e->name, op, from, election_owner);
+                     e->name, vote.op, vote.from, vote.election_owner);
             return election_error;
         }
-
-        op_copy = strdup(op);
-        uname_copy = strdup(from);
-
-        /* update the list of nodes that have voted */
-        g_hash_table_replace(e->voted, uname_copy, op_copy);
+        record_vote(e, &vote);
         reason = "Recorded";
         done = TRUE;
 
     } else {
-        struct timeval your_age;
-        const char *your_version = crm_element_value(vote, F_CRM_VERSION);
-        int tv_sec = 0;
-        int tv_usec = 0;
-
-        crm_element_value_int(vote, F_CRM_ELECTION_AGE_S, &tv_sec);
-        crm_element_value_int(vote, F_CRM_ELECTION_AGE_US, &tv_usec);
-
-        your_age.tv_sec = tv_sec;
-        your_age.tv_usec = tv_usec;
-
-        age = crm_compare_age(your_age);
-        if (crm_str_eq(from, e->uname, TRUE)) {
-            char *op_copy = NULL;
-            char *uname_copy = NULL;
-
-            if (!we_are_owner) {
-                crm_warn("Cannot count our own %s %s because we are not election owner (%s)",
-                         e->name, op, election_owner);
-                return election_error;
-            }
-            op_copy = strdup(op);
-            uname_copy = strdup(from);
-
-            /* update ourselves in the list of nodes that have voted */
-            g_hash_table_replace(e->voted, uname_copy, op_copy);
-            reason = "Recorded";
-            done = TRUE;
+        // A peer vote requires a comparison to determine which node is better
+        int age = crm_compare_age(vote.age);
+        int version = compare_version(vote.version, CRM_FEATURE_SET);
```

Comment:
Would rename the variable for greater clarity.

Reply:
will do